File: algorithms/423. Reconstruct Original Digits from English/solution.py
class Solution(object):
    def originalDigits(self, s):
        """
        :type s: str
        :rtype: str
        """
        result = [''] * 26
        counter = collections.Counter(s)
        unique = [['z', '0', 'zero'], ['w', '2', 'two'],
                  ['u', '4', 'four'], ['o', '1', 'one'],
                  ['r', '3', 'three'], ['f', '5', 'five'],
                  ['x', '6', 'six'], ['s', '7', 'seven'],
                  ['g', '8', 'eight'], ['i', '9', 'nine']]
        for c, d, w in unique:
            n = counter[c]
            result[ord(d) - ord('0')] = d * n
            for cc in w:
                counter[cc] -= n
        return ''.join(result)

File: algorithms/423. Reconstruct Original Digits from English/solution.cc
class Solution {
public:
    string originalDigits(string s) {
        vector<int> nums(10);
        vector<int> indices = { 0, 2, 4, 1, 3, 5, 6, 7, 8, 9 };
        vector<char> unique = { 'z', 'w', 'u', 'o', 'r', 'f', 'x', 's', 'g', 'i' };
        vector<string> words = { "zero", "two", "four", "one", "three", "five", "six", "seven", "eight", "nine" };
        unordered_map<char, int> counter;
        for ( char c : s ) {
            ++counter[c];
        }
        for ( int i = 0; i < 10; ++i ) {
            nums[indices[i]] = counter[unique[i]];
            int n = counter[unique[i]];
            for ( char c : words[i] ) {
                counter[c] -= n;
            }
        }
        string result;
        for ( int i = 0; i < 10; ++i ) {
            result += string(nums[i], i + '0');
        }
        return result;
    }
};
